model/modeldecoder/rumv3/decoder_test.go

```diff
@@ -0,0 +1,109 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package rumv3
+
+import (
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/model/modeldecoder/modeldecodertest"
+)
+
+func TestResetModelOnRelease(t *testing.T) {
+	inp := `{"m":{"se":{"n":"service-a"}}}`
+	m := fetchMetadataRoot()
+	require.NoError(t, decoder.NewJSONIteratorDecoder(strings.NewReader(inp)).Decode(m))
+	require.True(t, m.IsSet())
+	releaseMetadataRoot(m)
+	assert.False(t, m.IsSet())
+}
+
+func TestDecodeNestedMetadata(t *testing.T) {
+	t.Run("decode", func(t *testing.T) {
+		var out model.Metadata
+		testMinValidMetadata := `{"m":{"se":{"n":"name","a":{"n":"go","ve":"1.0.0"}}}}`
+		dec := decoder.NewJSONIteratorDecoder(strings.NewReader(testMinValidMetadata))
+		require.NoError(t, DecodeNestedMetadata(dec, &out))
+		assert.Equal(t, model.Metadata{Service: model.Service{
+			Name:  "name",
+			Agent: model.Agent{Name: "go", Version: "1.0.0"}}}, out)
+
+		err := DecodeNestedMetadata(decoder.NewJSONIteratorDecoder(strings.NewReader(`malformed`)), &out)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "decode")
+	})
+
+	t.Run("validate", func(t *testing.T) {
+		inp := `{}`
+		var out model.Metadata
+		err := DecodeNestedMetadata(decoder.NewJSONIteratorDecoder(strings.NewReader(inp)), &out)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "validation")
+	})
+
+}
+
+func TestMappingToModel(t *testing.T) {
+	expected := func(s string) model.Metadata {
+		return model.Metadata{
+			Service: model.Service{Name: s, Version: s, Environment: s,
+				Agent:     model.Agent{Name: s, Version: s},
+				Language:  model.Language{Name: s, Version: s},
+				Runtime:   model.Runtime{Name: s, Version: s},
+				Framework: model.Framework{Name: s, Version: s}},
+			User:   model.User{Name: s, Email: s, ID: s},
+			Labels: common.MapStr{s: s},
+		}
+	}
+
+	// setup:
+	// create initialized modeldecoder and empty model metadata
+	// map modeldecoder to model metadata and manually set
+	// enhanced data that are never set by the modeldecoder
+	val := reflect.New(reflect.TypeOf(metadata{}))
+	modeldecodertest.SetStructValues(val, "init", 5000)
+	m := val.Interface().(*metadata)
+	var modelM model.Metadata
+	mapToMetadataModel(m, &modelM)
```

Comment:
```suggestion
	var m metadata
	modeldecodertest.SetStructValues(&m, "init", 5000)
	var modelM model.Metadata
	mapToMetadataModel(&m, &modelM)
```

(illustration of changed API)